### ISSUE TITLE
Using debounce in renderEvents

### ISSUE THREAD

**Issue by melwas** [fullcalendar/fullcalendar#2938]
Hi, 

Have you thought about using debounce function when rendering events? I did a really quick test and replaced `renderEvents` with debounced version (in `reportEvents`) and performance improvement is significant (e.g. when adding and removing eventSources).

``` javascript
var renderEventsDebounced = debounce(renderEvents, 50);
```

I am just curious, because maybe you have tried this before but discovered some important disadvantages?

**Comment by arshaw** [comment]
this is a route i want to go down. this is a solution to #2681 an other issues where people want to avoid needless event rerendering. the downside is that event rendering becomes asynchronous, so it might make it difficult for people to know when it's finished, but that's where callbacks like [eventAfterAllRender](http://fullcalendar.io/docs/event_rendering/eventAfterAllRender/) come in.

**Comment by laoneo** [comment]
+1

**Comment by arshaw** [comment]
would help with "performance issues" #2524

**Comment by M0nter0** [comment]
Thanks a lot, @melwas 
It's a great solution. In my case, we were adding 100 slots at once, freezing the browser. Now using debounce() with 50 miliseconds they appear "instantly" without problems.

**Comment by kferrio** [comment]
+1

**Comment by rwwagner90** [comment]
Would love to be able to use this solution, could someone point me to where I need to put this debounce in? Is it here? https://github.com/fullcalendar/fullcalendar/blob/master/src/Calendar.js#L274

**Comment by arshaw** [comment]
you'd likely want to debounce the following lines: https://github.com/fullcalendar/fullcalendar/blob/v2.9.1/src/Calendar.js#L792-L794

**Comment by lunayo** [comment]
how should I change the code? I tried this but the performance is still quite slow
```
function renderEvents() { // destroys old events if previously rendered
  if (elementVisible()) {
	//freezeContentHeight();
	//currentView.displayEvents(events);
	//unfreezeContentHeight();
	debounce(currentView.displayEvents(events), 50);
   }
}
```

**Comment by jamesdixon** [comment]
@arshaw is this on the roadmap for a 2.x release or are you looking at overall performance improvements being part of 3.x?

**Comment by arshaw** [comment]
@lunayo you're using debounce incorrectly. it should be something like this:
```js
var debouncedDisplayEvents = debounce(function() {
	freezeContentHeight();
	currentView.displayEvents(events);
	unfreezeContentHeight();
}, 50);

function renderEvents() { // destroys old events if previously rendered
	if (elementVisible()) {
		debouncedDisplayEvents()
	}
}
```

@jamesdixon I'm thinking about it because introducing debounce will likely be a breaking change, and I might want to get it into a major release while I can. Though I can think of other ways to phase it in too.

**Comment by rwwagner90** [comment]
@arshaw could have some sort of flag to set for it. Then people could toggle it on and off.

**Comment by lunayo** [comment]
@arshaw I have tried the code but the performance still has no improvement. I have been using ajax `function events` to load thousands of events and return all the events using callback function. Should I change the event loading method to addEventSource instead?

**Comment by Manuweb2** [comment]
Hi @arshaw,

Thanks for the amazing job on fullcalendar.
We also noticed that using more than 600 events on a week calendar (yes we have clients with buzy weeks...) it took the calendar about 4s second to display the elements.
We are using a single source with an ajax call for the source using event: url and the ajax return a json.
The ajax call is returned in about 200ms, then we can count to 4s until the elements appear on the calendar...  we were using the 2.0.0beta until now and we thought updating to 3 would change that but it has no effect and still 4s until the elements appear. 
Just discovered this post and tryed the debounce method but no effect on performance.
When less that 100 elements it appears in less thant 100ms which is great.
We are using chrome and safari in our tests.
Are you thinking working on performance on the next 3.0.1 releases ? 

Thanks for the update, and congrats for the good work.
Manu

**Comment by arshaw** [comment]
implemented and released in [v3.1.0](https://github.com/fullcalendar/fullcalendar/releases/tag/v3.1.0)

`eventRenderWait`
https://fullcalendar.io/docs/event_rendering/eventRenderWait/

off by default.
